internal/util.go

```diff
@@ -0,0 +1,171 @@
+		value, err = parseValue(v.Value)
+	}
+	if err != nil {
+		// Should never happen.
+		return -1.0, err
+	}
+	return value, err
+}
+
+func parseValue(bytes []byte) (float64, error) {
+	value, err := strconv.ParseFloat(string(bytes), 64)
+	if err != nil {
+		return -1.0, fmt.Errorf("failed to parse value as float;value:<%s>", bytes)
+	}
+	return value, nil
+}
+
+func CreateMetricsList(c config.Config) ([]JsonMetric, error) {
+	var metrics []JsonMetric
+	for _, metric := range c.Metrics {
+		switch metric.Type {
+		case config.ValueScrape:
+			constLabels := make(map[string]string)
+			var variableLabels, variableLabelsValues []string
+			for k, v := range metric.Labels {
+				if len(v) < 1 || v[0] != '$' {
+					// Static value
+					constLabels[k] = v
+				} else {
+					variableLabels = append(variableLabels, k)
+					variableLabelsValues = append(variableLabelsValues, v)
+				}
+			}
+			jsonMetric := JsonMetric{
+				Desc: prometheus.NewDesc(
+					metric.Name,
+					metric.Help,
+					variableLabels,
+					constLabels,
+				),
+				KeyJsonPath:     metric.Path,
+				LabelsJsonPaths: variableLabelsValues,
+			}
+			metrics = append(metrics, jsonMetric)
+		case config.ObjectScrape:
+			for subName, valuePath := range metric.Values {
+				name := MakeMetricName(metric.Name, subName)
+				constLabels := make(map[string]string)
+				var variableLabels, variableLabelsValues []string
+				for k, v := range metric.Labels {
+					if len(v) < 1 || v[0] != '$' {
+						// Static value
+						constLabels[k] = v
+					} else {
+						variableLabels = append(variableLabels, k)
+						variableLabelsValues = append(variableLabelsValues, v)
+					}
+				}
+				jsonMetric := JsonMetric{
+					Desc: prometheus.NewDesc(
+						name,
+						metric.Help,
+						variableLabels,
+						constLabels,
+					),
+					KeyJsonPath:     metric.Path,
+					ValueJsonPath:   valuePath,
+					LabelsJsonPaths: variableLabelsValues,
+				}
+				metrics = append(metrics, jsonMetric)
+			}
+		default:
+			return nil, fmt.Errorf("Unknown metric type: '%s', for metric: '%s'", metric.Type, metric.Name)
+		}
+	}
+	return metrics, nil
+}
+
+func FetchJson(ctx context.Context, logger log.Logger, endpoint string, config config.Config) ([]byte, error) {
+	httpClientConfig := config.HTTPClientConfig
+	client, err := pconfig.NewClientFromConfig(httpClientConfig, "fetch_json", true)
+	if err != nil {
+		level.Error(logger).Log("msg", "Error generating HTTP client", "err", err) //nolint:errcheck
+		return nil, err
+	}
+	req, err := http.NewRequest("GET", endpoint, nil)
+	req = req.WithContext(ctx)
+	if err != nil {
+		level.Error(logger).Log("msg", "Failed to create request", "err", err) //nolint:errcheck
+		return nil, err
+	}
+
+	for key, value := range config.Headers {
+		req.Header.Add(key, value)
+	}
+	if req.Header.Get("Accept") == "" {
+		req.Header.Add("Accept", "application/json")
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
```

Comment:
If there is a 400, we would not reuse connections unless we discard the body.
```suggestion
	defer func(){
	io.Copy(ioutil.Discard, resp.Body)
	resp.Body.Close()
	}
```

From the documentation:
> The default HTTP client's Transport may not
> reuse HTTP/1.x "keep-alive" TCP connections if the Body is
> not read to completion and closed.


In this case, if resp code is not 200, we do not exhause the body, so we need this.

Reply:
Fixed!